File: a2sv_contest_problems/M_Moving_Both_Hands.cpp
#include <string>
#include <vector>
#include <iostream>
#include <deque>
#include <tuple>
#include <algorithm>
#include <map>
#include <bits/stdc++.h>

#define ll long long

using namespace std;

int main()
{
    ios::sync_with_stdio(false);
    cin.tie(nullptr);

    ll n, m;
    cin >> n >> m;

    map<ll, vector<pair<ll, ll>>> graph;
    // reversed graph
    map<ll, vector<pair<ll, ll>>> r_graph;

    vector<vector<ll>> best(2, vector<ll>(n + 1, __LONG_LONG_MAX__));

    best[0][1] = 0;
    best[1][1] = 0;

    for (auto i = 0; i < m; ++i)
    {
        ll u, v, w;
        cin >> u >> v >> w;

        graph[u].push_back({v, w});
        r_graph[v].push_back({u, w});
    }

    priority_queue<tuple<ll, ll, ll>> heap;
    heap.push({0, 1, 0});

    while (!heap.empty())
    {
        auto [distance, node, direction] = heap.top();
        distance *= -1;
        heap.pop();

        if (distance > best[direction][node])
            continue;

        if (direction == 0)
        {

            for (auto [nei, w] : graph[node])
            {
                if (w + distance < best[0][nei])
                {
                    heap.push({-1 * (w + distance), nei, 0});
                    best[0][nei] = w + distance;
                }
            }
        }

        for (auto [nei, w] : r_graph[node])
        {
            if (w + distance < best[1][nei])
            {
                heap.push({-1 * (w + distance), nei, 1});
                best[1][nei] = w + distance;
            }
        }
    }

    for (auto i = 2; i <= n; ++i)
    {
        auto ans = min(best[0][i], best[1][i]);
        if (ans == __LONG_LONG_MAX__)
            ans = -1;
        cout << ans << (i != n ? ' ' : '\n');
    }

    return 0;
}

File: a2sv_contest_problems/M_Moving_Both_Hands.py
from math import inf
from collections import defaultdict
from heapq import heappop, heappush
from sys import stdin, stdout

n, m = map(int, stdin.readline().split())

graph = defaultdict(list)
reverse_graph = defaultdict(list)

for _ in range(m):
    u, v, w = map(int, stdin.readline().split())
    graph[u].append((v, w))
    reverse_graph[v].append((u, w))


best = [[inf for _ in range(n + 1)] for _ in range(2)]

heap = [(0, 1, 0)]

while heap:
    distance, node, direction = heappop(heap)
    if distance > best[direction][node]:
        continue

    if direction == 0:
        for (nei, w) in graph[node]:
            if w + distance < best[direction][nei]:
                heappush(heap, (w + distance, nei, direction))
                best[direction][nei] = w + distance

    for (nei, w) in reverse_graph[node]:
        if w + distance < best[1][nei]:
            heappush(heap, (w + distance, nei, 1))
            best[1][nei] = w + distance


print(*map(lambda x: min(best[0][x], best[1][x])
      if min(best[0][x], best[1][x]) != inf else -1, range(2, n + 1)))
